### PR TITLE
add linting support for script folder

### PR DESCRIPTION
**Pull request by shiftkey**
Currently, when you try and package from `master`, you'll get this error:

```shellsession
~/src/desktop/ yarn run package
yarn run v1.3.2
$ ts-node script/package.ts

/Users/shiftkey/src/desktop/node_modules/ts-node/src/index.ts:307
        throw new TSError(formatDiagnostics(diagnosticList, cwd, ts, lineOffset))
              ^
TSError: ⨯ Unable to compile TypeScript
script/package.ts (126,20): Cannot find name 'ElectronInstallerRedhat'. (2304)
script/package.ts (136,24): Parameter 'err' implicitly has an 'any' type. (7006)
script/package.ts (147,20): Cannot find name 'ElectronInstallerDebian'. (2304)
script/package.ts (157,24): Parameter 'err' implicitly has an 'any' type. (7006)
script/package.ts (181,20): Cannot find name 'ElectronInstallerAppImage'. (2304)
    at getOutput (/Users/shiftkey/src/desktop/node_modules/ts-node/src/index.ts:307:15)
    at /Users/shiftkey/src/desktop/node_modules/ts-node/src/index.ts:336:16
    at Object.compile (/Users/shiftkey/src/desktop/node_modules/ts-node/src/index.ts:498:11)
    at Module.m._compile (/Users/shiftkey/src/desktop/node_modules/ts-node/src/index.ts:392:43)
    at Module._extensions..js (module.js:623:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/Users/shiftkey/src/desktop/node_modules/ts-node/src/index.ts:395:12)
    at Module.load (module.js:531:32)
    at tryModuleLoad (module.js:494:12)
    at Function.Module._load (module.js:486:3)
    at Function.Module.runMain (module.js:653:10)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

You'll also see this when you open the file in VSCode.

screen shot 2017-11-04 at 9 17 07 am

I'm not quite sure when exactly this was introduced, but it doesn't really matter.

We should treat these the same as our other source code. So I'm adding a `compile:script` step to run alongside our linting tasks, and a `tsconfig.json` to ensure we're emitting the right sort of JS to be executed by Node.

The presence of the `tsconfig.json` is the fix for the editor, the rest is just automation to catch regressions on CI.

### CODE REVIEW

File: script/eslint.ts [hunk 1]
```diff
@@ -26,7 +26,8 @@ const client: IClient = require('eslint_d/lib/client')
 const ESLINT_ARGS = [
   '--cache',
   '--rulesdir=./eslint-rules',
-  './{script,eslint-rules}/**/*.{j,t}s?(x)',
+  './eslint-rules/**/*.{j,t}s?(x)',
+  './script/**/*.ts',
```
Comment (shiftkey) [inline script/eslint.ts]: These have been split up to ensure the generated JS files under `script/` aren't linted.

File: script/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    "moduleResolution": "node",
+    "target": "es6",
+    "noImplicitAny": true,
+    "noImplicitReturns": true,
+    "noImplicitThis": true,
+    "noFallthroughCasesInSwitch": true,
+    "noUnusedLocals": true,
+    "strictNullChecks": true,
+    "sourceMap": false,
+    "noEmit": true
```
Comment (shiftkey) [inline script/tsconfig.json]: Pro Tip: this is how you prevent `tsc` from emitting the compiled JS

File: script/eslint.ts (hunk 1 above)
Comment (shiftkey) [reply script/eslint.ts]: Nevermind, this isn't necessary any more.

File: package.json [hunk 3]
```diff
@@ -19,12 +19,13 @@
...
     "build:dev": "yarn compile:dev && cross-env NODE_ENV=development ts-node script/build.ts",
     "build:prod": "yarn compile:prod && cross-env NODE_ENV=production ts-node script/build.ts",
-    "package": "ts-node script/package.ts",
+    "package": "ts-node -P script script/package.ts",
     "generate-octicons": "ts-node script/generate-octicons.ts",
     "clean:tslint": "rimraf tslint-rules/*.js",
     "compile:tslint": "tsc -p tslint-rules",
-    "lint": "yarn compile:tslint && yarn tslint && yarn eslint-check && yarn eslint",
-    "tslint": "tslint ./{scripts,tslint-rules}/*.ts ./app/{src,typings,test}/**/*.{ts,tsx}",
+    "compile:script": "tsc -p script",
```
Comment (joshaber) [inline package.json]: Just to make sure I understand, this is really to type check the scripts? `compile` makes me think there's actually gonna be output. What do you think about changing the name?

Comment (shiftkey) [reply package.json]: Yeah, this is the tricky part.

We use `ts-node` when invoking our scripts so we don't need to emit the JS files to disk, and I don't see the value in changing that. But we also want to catch this before we have to invoke the script, so "compiling" felt like the closest term here (and we're leveraging `tsc` here so I'm somewhat keen to stick to that word).

If there's a more appropriate name here ("lint" doesn't feel right, and we already have a lot of linting going on) then I'm fine with renaming it.

**Comment by aa369888** [comment]
我現在應該從哪裡開始學習

2017年11月7日 07:40，"Brendan Forster" <notifications@github.com>寫道：

File: package.json (hunk 3 above)
Comment (joshaber) [reply package.json]: Eh, I'll refrain from bike shedding it.

**Comment by aa369888** [comment]
我能夠知道你們在做些什麼嗎

**Comment by aa369888** [comment]
能不能讓我知道

2017年11月7日 08:10，"aa369888" <notifications@github.com>寫道：

**Pull request merged**